tests/Test/Xrefcheck/TooManyRequestsSpec.hs

```diff
@@ -149,35 +158,118 @@ test_tooManyRequests = testGroup "429 response tests"
       E.bracket (forkIO $ mock429WithGlobalIORef infoReverseAccumulatorRef) killThread $ \_ -> do
-        _ <- verifyLink "http://127.0.0.1:5000/429grandfinale"
+        _ <- verifyLinkDefault setRef "http://127.0.0.1:5000/429grandfinale"
         infoReverseAccumulator <- readIORef infoReverseAccumulatorRef
         reverse infoReverseAccumulator @?=
           [ ("HEAD", tooManyRequests429)
           , ("HEAD", serviceUnavailable503)
           , ("GET", ok200)
           ]
+    , testGroup "Timeout tests"
+      [ testCase "Returns an error on timeout if there were no 429 responses" $
+          timeoutTestCase [Delay] False
+            (cNetworkingL . ncMaxTimeoutRetriesL .~ 2)
+      , testCase "Returns an error if there were 429 but no timeouts allowed" $
+          timeoutTestCase [Respond429, Delay] False
+            (cNetworkingL . ncMaxTimeoutRetriesL .~ 0)
+      , testCase "Succeeds if there were 429 and one timeout allowed" $
+          timeoutTestCase [Respond429, Delay] True
+            (cNetworkingL . ncMaxTimeoutRetriesL .~ 1)
+      , testCase "Fails on second timeout if there were 429 and one timeout allowed" $
+          timeoutTestCase [Respond429, Delay, Delay] False
+            (cNetworkingL . ncMaxTimeoutRetriesL .~ 1)
+      , testCase "Fails on second timeout if there were 429 and one timeout allowed" $
+          timeoutTestCase [Respond429, Delay, Delay] False
+            (cNetworkingL . ncMaxTimeoutRetriesL .~ 1)
+      , testCase "Fails on maximum allowed errors achieved (mixed errors)" $
+          timeoutTestCase [Respond429, Delay, Respond429, Delay] False $ \c -> c
+            & cNetworkingL . ncMaxTimeoutRetriesL .~ 3
+            & cNetworkingL . ncMaxRetriesL .~ 3
+      , testCase "Fails on timeout if another domain returned 429" $ do
+          setRef <- newIORef S.empty
+          checkMultipleLinksWithServer
+            (mockTimeout (sec 0.4) [Respond429, Ok, Delay])
+            setRef
+            [ VerifyLinkTestEntry{ vlteConfigModifier = \c -> c
+                                     & cNetworkingL . ncMaxRetriesL .~ 1
+                                     & setAllowedTimeout
+                                 , vlteLink = "http://127.0.0.1:5000/timeout"
+                                 , vlteProgress = mkProgressWithOneTask True
+                                 , vlteExpectation = VerifyResult []
+                                 }
+            , VerifyLinkTestEntry{ vlteConfigModifier = \c -> c
+                                     & cNetworkingL . ncMaxTimeoutRetriesL .~ 0
+                                     & setAllowedTimeout
+                                 , vlteLink = "http://localhost:5000/timeout"
+                                 , vlteProgress = mkProgressWithOneTask False
+                                 , vlteExpectation = VerifyResult
+                                   [ ExternalHttpTimeout (Just $ DomainName "localhost")
+                                   ]
```

Comment:
Earlier you commented:
> At least one test I would like to also see: involve two sites, and test that 429 in one and a timeout in the other is treated respectively; currently this is not covered at all.

Doesn't this mean what I've added? This test checks that returning 429 from one site doesn't affect treating timeouts for another site.

Reply:
> This test checks that returning 429 from one site doesn't affect treating timeouts for another site

This is a thing that is also worth testing, but we had this behaviour before our new feature too - previously timeout always resulted in `Response timeout` error, and here you test that we get `Response timeout` in case of 429. This test does not check that the feature has changed something.

I rather meant to test the opposite thing: that 429 will prevent the next `Response timeout` if config allows it.